File: Classic Algorithms/collatz.js
function collatz_iter(num){
	var steps = 0;
	while (num  > 1){
		if (num % 2 === 0) {
			num = num / 2;
			steps += 1;
		} else {
			num = 3*num + 1;
			steps += 1
		}
	}
	return steps;
}

function collatz_recur(num,steps){
	if (num <= 1) {
		return steps;
	}
	else{
		if (num % 2 === 0) {
			return collatz_recur(num / 2,steps+1);
		} else {
			return collatz_recur(3*num+1,steps+1);
		}
	};

}

var collatz_memo = function(){
	this.memo = {};
	that = this;
	this.collatz = function (num){
		//Have we seen this initial argument before?
		var steps = that.memo[num];
		if (steps) {
			return steps
		} else {
			steps = 0
		}

		var orig_num = num;
		while (num > 1){
			//Have we seen this intermediate number before?
			var prev_steps = that.memo[num]
			if (prev_steps) {
				that.memo[orig_num] = steps + prev_steps;
				return steps + prev_steps;
			};
			steps += 1;
			if (num % 2 === 0) {
				num = num / 2;
			} else {
				num = 3*num + 1;
			}
		}
		that.memo[orig_num] = steps;
		return steps;
	}
	return that;
}

var elapsed_time = function(note){
    var precision = 3; // 3 decimal places
    var elapsed = process.hrtime(start)[1] / 1000000; // divide by a million to get nano to milli
    console.log(process.hrtime(start)[0] + " s, " + elapsed.toFixed(precision) + " ms - " + note); // print message + time
    start = process.hrtime(); // reset the timer
}
var limit = 2000;
var numbers = []
for (var i = 0; i < limit; i++) {
	numbers.push(i);
};
var start = process.hrtime();
elapsed_time("begin")
for (var i = 0; i < numbers.length; i++) {
	collatz_iter(numbers[i])
};
elapsed_time("iter time")
for (var i = 0; i < numbers.length; i++) {
	collatz_recur(numbers[i],0)
};
elapsed_time("recur time")
var obj = collatz_memo();
for (var i = 0; i < numbers.length; i++) {
	obj.collatz(numbers[i])
};
elapsed_time("memo time")

File: Classic Algorithms/collatz.py
#collatz.py
import time

class collatz_memo:
	def __init__(self):
		self.memo = {}

	def collatz(self,num):
		steps = self.memo.get(num,0)
		if steps != 0:
			return self.memo[num]
		orig_num = num
		while num != 1:
			prev_steps = self.memo.get(num,0)
			if prev_steps != 0:
				# print "other memo hit"
				self.memo[orig_num] = steps + prev_steps
				return steps + prev_steps
			steps += 1
			if num % 2 == 0:
				num = num / 2
			else:
				num = 3 * num + 1
		self.memo[orig_num] = steps
		return steps



def collatz_iter(num):
	steps = 0;
	while num != 1:
		if num % 2 == 0:
			num = num /2
			steps += 1
		else:
			num = num * 3 + 1
			steps += 1
	return steps

def collatz_recur(num,steps):
	if num == 1:
		return steps
	elif num % 2 == 0:
		return collatz_recur(num/2,steps+1)
	else:
		return collatz_recur(3*num+1,steps+1)

def millis():
	return int(round(time.time()*1000))

def elapsed_time(note,start):
	elapsed = millis() - start
	print "%s in %f" % (note, elapsed)
	start = millis()

if __name__ == '__main__':
	limit = 2000
	start = time.clock()
	obj = collatz_memo()
	answers1 = [obj.collatz(x) for x in range(1,limit)]
	end1 = time.clock()
	answers2 = [collatz_iter(x) for x in range(1,limit)]
	end2 = time.clock()
	answers3 = [collatz_recur(x,0) for x in range(1,limit)]
	end3 = time.clock()
	print "memo took %f" % ((end1 - start)*1000)
	print "iter took %f" %  ((end2 - end1)*1000)
	print "recu took %f" % ((end3 - end2)*1000)
	# start = millis()
	# print collatz_iter(50000000000)
	# elapsed_time("iter done",start)
	# print collatz_recur(50000000000, 0)
	# elapsed_time("recur done",start)
	